### PR TITLE
Refactor plugin installer and account provider

### PR DESCRIPTION
**Pull request by aiyengar2**
This PR performs a refactor of the codebase to make it easier to add tests and adds a ton of tests.

On a high-level, the commits are organized into the following "phases":

### Misc. Changes

The purpose of this phase is to just add in general improvements to CI or testing that can go in the current code.

- Add scripts/generate-coverage: a simple script that allows you to generate HTML-based Go coverage
- Add simple test for pkg/version for coverage
- Create pkg with utils that are only used for tests

### Defining a central utility package

The purpose of this phase is to centralize all operations that are done throughout the codebase using the `os` and `exec` packages into a single utility module.

This utility module contains all hard-coded paths (i.e. directories and UIDs), allows you to execute or "dry-execute" Powershell scripts (if running in DryRun mode or running on a non-Windows host) and introduces a "filesystem manager" based on https://github.com/go-git/go-billy that allows tests to mock the host filesystem using an in-memory filesystem to efficiently run *most* tests.

There are also other small common utility functions that will be introduced in this module, such as the helper function that validates namespaces.

- Add constant paths to new pkg/utils
- Add filesystem manager to pkg/utils
- Add filesystem functions to pkg/utils
- Add ability to execute powershell scripts
- Add simple helper function to validate namespace to pkg/utils

### Refactor Plugin Manager / Installer

The purpose of this phase is to refactor the plugin manager (now called the installer).

On a high-level, the biggest difference in the new installer is that it follows a two stage process: 1) It figures out the current status of the DLL install (using a new status module) and 2) it actually executes only the steps that need to be done.

Also, this version of the installer can run on non-Windows systems but will effectively only move around files and dry-run the execution of powershell scripts that register the DLL.

The only enhancement added to the Plugin Manager is the ability to specify a timeout environment variable to installer operations.

- Rename pkg/plugin/manager -> pkg/installer
- Store all embedded files in a single module
- Add status module for installer
- Rename installer files to include building them on nix systems
- Refactor core logic in installer to utilize new helper modules
- Add timeout env var to installer operations

### Refactor Account Provider

The purpose of this phase is to refactor the account provider.

On a high-level, this PR replaces a lot of the old norman code with a wrangler controller that watches secrets in the cluster, which means that the Account Provider no longer has to make a client call to the Kubernetes API every time that it needs to refresh its value of a Kubernetes secret; instead, it just responds with what is within its cache.

In addition, this PR does a core refactor of the way that the server logic was defined to make it easier to test; it moves the https://github.com/gorilla/mux logic into a simple NewHandler function call (which means that there is no longer a hard dependency on gorilla/mux as an HTTP engine) and refactors the server to no longer handle creating the port file automatically.

- Rename pkg/plugin/provider -> pkg/provider
- Update default namespace to watch for Secrets to cattle-windows-gmsa-system
- Replace norman generated code with wrangler controller and utilize controller cache instead of client
- Move Response struct into server.go
- Remove client struct entirely in favor of directly starting controllers from main.go
- Move all debug logic into main
- Move the core handler logic into a new server module
- Move server.go into pkg/provider/server
- Add new manager module to handle creating and importing certificates using functions from util module
- Refactor core functionality to use new helper modules
- Register at least one controller so that deferred caches start
- Update .gitignore and README.md

---

On a higher level, the only core changes in behavior that should be observed by a user with this refactor are:
1. Docs are added for manual certificate rotation.
2. A new timeout env var was introduced to the DLL installer.

Related Issue: https://github.com/rancher/Rancher-Plugin-gMSA/issues/24

### CODE REVIEW

File: pkg/utils/filesystem.go [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+		if err := DeleteFile(tempFile); err != nil {
+			return fmt.Errorf("unable to delete temporary file %s: %s", tempFile, err)
+		}
+		return nil
+	}
+	renameFunc := os.Rename
+	if IsMemFs() {
+		// in a dry-run scenario, running os.Remove would fail so this mimics that process
+		// Note: the reason why we use os.Remove in the real-life scenario is because that is
+		// how Microsoft recommends performing DLL Upgrades
```
[the remark this answers is not in the document]
Comment (HarrisonWAffel) [reply pkg/utils/filesystem.go:216]: (unless this is going to be removed in relation to other certs work)

**Review by HarrisonWAffel — CHANGES_REQUESTED** [review]
Generally looks good, just some nits here and there

File: pkg/provider/manager/manager.go [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+	"path/filepath"
+	"sync"
+
+	"github.com/aiyengar2/Rancher-Plugin-gMSA/pkg/utils"
+	"github.com/sirupsen/logrus"
+	"go.uber.org/multierr"
+)
+
+var (
+	UserProvidedCertificates = map[string][]string{
```
[the remark this answers is not in the document]
Comment (aiyengar2) [reply pkg/provider/manager/manager.go:15]: Makes sense, renaming this `Certificates`

File: pkg/provider/server/server.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+		return nil, fmt.Errorf("failed to identify port from net listener: %s", err)
+	}
+	logrus.Infof("Listening on port %d", h.port)
+	return ln, nil
+}
+
+func (h *HTTPServer) serve(ctx context.Context, ln net.Listener) {
+	server := http.Server{
+		Handler:     h.Handler,
+		BaseContext: func(net.Listener) context.Context { return ctx },
```
[the remark this answers is not in the document]
Comment (HarrisonWAffel) [reply pkg/provider/server/server.go:53]: Overlooked that this was a `http` struct and not a custom type from this project, this makes sense now

File: pkg/utils/exec.go [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+	cmd := fmt.Sprintf(`certutil %s | Select-String -Pattern "Cert Hash\(sha1\):*"`, certPath)
+	out, err := RunPowershellCommandWithOutput(cmd)
+	if err != nil {
+		return fmt.Errorf("could not find certificate: %s", err)
+	}
+	var thumbprint string
+	if out == nil {
+		thumbprint = "<THUMBPRINT>"
+	} else {
+		// Powershell was not actually executed
```
[the remark this answers is not in the document]
Comment (aiyengar2) [reply pkg/utils/exec.go:66]: moving comment 👍

File: pkg/provider/server/server.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (aiyengar2) [reply pkg/provider/server/server.go:53]: BaseContext takes in a function

**Review by HarrisonWAffel — APPROVED** [review]
Nice!

**Pull request merged**